bluepyefe/auto_targets.py

```diff
@@ -0,0 +1,182 @@
+     and efeatures that match the data available in the recordings."""
+
+    def __init__(
+        self,
+        protocols,
+        amplitudes,
+        efeatures,
+        min_recordings_per_amplitude=10,
+        preferred_number_protocols=1,
+        tolerance=10,
+    ):
+        """
+        Extract efeatures.
+
+        Args:
+            protocols (list of str): list of eCodes, by order of priority.
+            amplitudes (list of int): list of amplitudes, expressed as
+                percentages of the rheobase, by order of priority
+            efeatures (list str): list of efeatures.
+            min_recordings_per_amplitude (int): minimum number of recordings
+                a amplitude should have to be considered as a target.
+            preferred_number_protocols (int): out of all the available
+                amplitudes, how many should be used.
+            tolerance (float): tolerance used to bin the efeatures around the
+                target amplitudes.
+        """
+
+        self.efeatures = efeatures
+        self.protocols = protocols
+        self.amplitudes = amplitudes
+
+        self.min_recordings_per_amplitude = min_recordings_per_amplitude
+        self.preferred_number_protocols = preferred_number_protocols
+        self.tolerance = tolerance
+
+        self.active_amplitudes = []
+        self.active_ecode = []
+
+    def select_ecode_and_amplitude(self, recordings):
+
+        for ecode in self.protocols:
+
+            available_amps = []
+            for r in recordings:
+                if r.protocol_name == ecode and r.amp_rel is not None:
+                    available_amps.append(r.amp_rel)
+
+            for amp in self.amplitudes:
+                n_available_rec = sum(
+                    abs(a - amp) < self.tolerance for a in available_amps
+                )
+                if n_available_rec >= self.min_recordings_per_amplitude:
+                    self.active_ecode.append(ecode)
+                    self.active_amplitudes.append(amp)
+
+        if (
+            self.active_ecode and
+            len(self.active_ecode) > self.preferred_number_protocols
+        ):
+            self.active_ecode = self.active_ecode[
+                :self.preferred_number_protocols]
+            self.active_amplitudes = self.active_amplitudes[
+                :self.preferred_number_protocols]
+
+    def is_valid(self):
+        return bool(self.active_amplitudes) and bool(self.active_ecode)
+
+    def generate_targets(self):
+
+        targets = []
+
+        for amp, protocol_name in zip(
+                self.active_amplitudes, self.active_ecode):
+            for efeature in self.efeatures:
+                targets.append({
+                    "efeature": efeature,
+                    "protocol": protocol_name,
+                    "amplitude": amp,
+                    "tolerance": self.tolerance,
+                })
+
+        return targets
```

Comment:
Optional: the itertools.product can be used here to remove one level of nesting.

```python
        targets = []
        for (amp, protocol_name), efeature in itertools.product(
            zip(self.active_amplitudes, self.active_ecode), self.efeatures
        ):
            targets.append({
                    "efeature": efeature,
                    "protocol": protocol_name,
                    "amplitude": amp,
                    "tolerance": self.tolerance,
                })

        return targets

```

Reply:
Neat trick !